internals/src/compact_size.rs

```diff
@@ -0,0 +1,310 @@
+/// ```
+pub fn encode(value: u64) -> ([u8; 9], usize) {
+    let mut buf = [0_u8; 9];
+    let size;
+
+    match value {
+        0..=0xFC => {
+            size = 1;
+            buf[0] = value as u8; // Cast ok because of match.
+        }
+        0xFD..=0xFFFF => {
+            size = 3;
+            let v = value as u16; // Cast ok because of match.
+            buf[0] = 0xFD;
+            buf[1..size].copy_from_slice(&v.to_le_bytes());
+        }
+        0x10000..=0xFFFFFFFF => {
+            size = 5;
+            let v = value as u32; // Cast ok because of match.
+            buf[0] = 0xFE;
+            buf[1..size].copy_from_slice(&v.to_le_bytes());
+        }
+        _ => {
+            size = 9;
+            let v = value;
+            buf[0] = 0xFF;
+            buf[1..size].copy_from_slice(&v.to_le_bytes());
+        }
+    }
+    (buf, size)
+}
+
+/// Gets the `CompactSize` encoded at the front of `slice`.
+///
+/// Does not modify `slice`, enforces invariant that `size < CompactSize::MAX_SIZE`.
+///
+/// # Returns
+///
+/// The `CompactSize` if one exists as well as the number of bytes used to encode it.
+///
+/// # Example
+///
+/// ```
+/// use bitcoin_internals::compact_size;
+///
+/// // Extra bytes are ignored.
+/// let (compact, _) = compact_size::decode(&[0x20, 0xAB, 0xCD]).unwrap();
+/// assert_eq!(compact, 32);
+/// ```
+pub fn decode(slice: &[u8]) -> Result<(u64, usize), DecodeError> {
+    if slice.is_empty() {
+        return Err(DecodeError::Empty);
+    }
+
+    let res = match slice[0] {
+        0xFF => {
+            const SIZE: usize = 9;
+            if slice.len() < SIZE {
+                return Err(TooShortError::new(slice, SIZE).into());
+            };
+
+            let mut bytes = [0_u8; SIZE - 1];
+            bytes.copy_from_slice(&slice[1..SIZE]);
+            (u64::from_le_bytes(bytes), SIZE)
+        }
+        0xFE => {
+            const SIZE: usize = 5;
+            if slice.len() < SIZE {
+                return Err(TooShortError::new(slice, SIZE).into());
+            };
+
+            let mut bytes = [0_u8; SIZE - 1];
+            bytes.copy_from_slice(&slice[1..SIZE]);
+            let v = u32::from_le_bytes(bytes);
+            (u64::from(v), SIZE)
+        }
+        0xFD => {
+            const SIZE: usize = 3;
+            if slice.len() < SIZE {
+                return Err(TooShortError::new(slice, SIZE).into());
+            };
+
+            let mut bytes = [0_u8; SIZE - 1];
+            bytes.copy_from_slice(&slice[1..SIZE]);
+            let v = u16::from_le_bytes(bytes);
+            (u64::from(v), SIZE)
+        }
+        n => (u64::from(n), 1),
+    };
+    Ok(res)
+}
+
+/// Error returned when decoding a `CompactSize`.
+///
+/// In other words the object being serialized is too big to go over the wire.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DecodeError {
+    /// Tried to decode a `CompactSize` from an empty slice.
+    Empty,
```

Comment:
`Empty` is literally an instance of `TooShort`, we don't need it.

Reply:
Ha! Conceptually yes but we don't have the expected length until we read the first byte.

(I removed `Empty` at one stage to find this out :)

Reply:
To make Kix's point more explicit, the "expected minimum length" should be 1.

As for whether we should combine the two variants, I could go either way. I guess the fewer variants the better, especially here where we only have two variants and we could eliminate the whole enum.

Reply:
> "expected minimum length" should be 1.

Ah, adding "minimum" indeed solves the problem.

Reply:
I relized this is actually a mute point, first I thought we could not have this module in `internals` because the error type is pubic if returned ever, but it turns out, with this PR we always call `ok` so we should just return an `Option` from `decode` also.

Reply:
I'd prefer `Result<_, ()>` to `Option`.

Reply:
Also FTR we could've just wrapped the error.

Reply:
> I'd prefer `Result<_, ()>` to `Option.`

To imply its an error case but we don't provide a specific error?

Reply:
Yes, also triggers the `must_use` warning.

Reply:
`internals` is unstable anyway, but I'm not opposed to it.